From 12b17f067278f79e0e578af65f49f9ca91a9b21b Mon Sep 17 00:00:00 2001
From: Hongxu Jia <hongxu.jia@windriver.com>
Date: Sat, 22 May 2021 19:54:50 +0800
Subject: [PATCH] anspassd.c: correct blank password checking

Since commit [1] applied in libgcrypt, rerun anspassd
failed with segment fault
[snip]
$ export ANSPASS_PATH="/tmp/.abc"
$ ./anspassd
Please enter a password for unlocking credentials
(max of 256 characters)
Password:
Token: F3F830A1CD0B81F2

$ ./anspassd
Segmentation fault (core dumped)
[snip]

Previously, while anspassd checking blank password for zero
prompt run, it called gcry_md_hash_buffer with input string '\0'
and input len 1. In libgcrypt, gcry_md_hash_buffer indirectly
called _gcry_md_block_write, but _gcry_md_block_write will
directly return without memory copying at condition `if (!inbuf)'.

In commit [1], libgcrypt changed the return condition from `if (!inbuf)'
to `if (inlen == 0)'
-  if (!inbuf)  --->   +  if (inlen == 0)
-    return;           +    return;

It caused _gcry_md_block_write run with memory operation and
trigger a segment fault to anspassd

Improve blank password checking, to input NULL with len 0, then
_gcry_md_block_write of libgcrypt will return without memory
copying at condition `if (inlen == 0)'

[1] https://git.gnupg.org/cgi-bin/gitweb.cgi?p=libgcrypt.git;a=commit;h=e76617cbab018dd8f41fd6b4ec6740b5303f7e13

Upstream-Status: Pending

Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>
---
 anspassd.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/anspassd.c b/anspassd.c
index 5f5dcbc..59e8044 100644
--- a/anspassd.c
+++ b/anspassd.c
@@ -340,7 +340,7 @@ int load_db1(const struct db1_meta *db1m)
 
 	/* Check blank password for zero prompt run */
 	gcry_md_hash_buffer(GCRY_MD_MD5, (void *)passwd_hash, (const void*)
-			'\0', 1);
+			NULL, 0);
 	load_gcrypt(passwd_hash, &db1.passwd, &db1.global, &dec_global);
 
 	/* Request passwd if password isn't blank */
-- 
2.27.0

